feat: add more deprovisioning logs

tzneal opened the pull request:
**Description**

Adds more visibility regarding consolidation decisions/

**How was this change tested?**
Deployed to EKS.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

tzneal commented:
Sample of New logs
```

karpenter-764d9bf778-jfgtd controller 2022-12-06T17:56:57.849Z	INFO	controller.deprovisioning	node ip-192-168-134-177.us-west-2.compute.internal can't be consolidated, can't replace with a cheaper node	{"commit": "5d20fe2"}


karpenter-764d9bf778-jfgtd controller 2022-12-06T17:59:19.151Z	INFO	controller.deprovisioning	node ip-192-168-134-177.us-west-2.compute.internal can't be consolidated, pdb default/inflate-pdb prevents pod evictions	{"commit": "5d20fe2"}


karpenter-ff5c78494-mkqdg controller 2022-12-06T18:40:01.118Z	INFO	controller.provisioner	pod default/inflate-anti-self-55894c5d8b-522jd has a preferred Anti-Affinity which can prevent consolidation	{"commit": "5d20fe2"}
karpenter-ff5c78494-mkqdg controller 2022-12-06T18:40:01.118Z	INFO	controller.provisioner	pod default/inflate-anti-self-55894c5d8b-wtkmt has a preferred Anti-Affinity which can prevent consolidation	{"commit": "5d20fe2"}

```

Sample of new Events
```
Events:
  Type     Reason                   Age                From             Message
  ----     ------                   ----               ----             -------
  Normal   Unconsolidatable         66s                karpenter        pdb default/inflate-pdb prevents pod evictions
```

tzneal commented on pkg/controllers/deprovisioning/reporter.go:
```diff
@@ -0,0 +1,41 @@
+	recorder events.Recorder
+}
+
+func NewReporter(recorder events.Recorder) *Reporter {
```
I pulled this out thinking it might be useful for reporting upgrade and drift as well.

ellistarn commented on pkg/utils/pretty/changemonitor.go:
```diff
@@ -34,6 +34,12 @@ func NewChangeMonitor() *ChangeMonitor {
 
+// Reconfigure allows reconfiguring the change monitor with a new duration. This resets any previously recorded
+// changes and should only be done at construction.
+func (c *ChangeMonitor) Reconfigure(expiration time.Duration) {
```
Why not pass this in the constructor as an option, rather than making it and then mutating it?

tzneal replied:
There were a bunch of other usages already. I figured I would leave 24h as the default and allow changing it post construction.

ellistarn reviewed (dismissed):
Neat!

tzneal commented:
Merge conflicts resolved.

tzneal commented:
@ellistarn I was debating on removing the logs from this and just leaving the events, WDYT?

ellistarn commented:
> @ellistarn I was debating on removing the logs from this and just leaving the events, WDYT?

I'd really love an abstraction = event+log for major things that we're doing. I've heard customers asking for events recently.

tzneal commented:
> > @ellistarn I was debating on removing the logs from this and just leaving the events, WDYT?
> 
> I'd really love an abstraction = event+log for major things that we're doing. I've heard customers asking for events recently.

The reason I had second thoughts here are these are things we are not doing for the most part (e.g. not consolidating node XYZ due to a PDB).  They'll repeat every hour, which is fine for an event, but I thought it might be too noisy for logs.

ellistarn commented:
I wonder if it gives the wrong impression if we only do it every hour, since we're checking much more frequently. Definitely agree that we don't want to spam on this.

tzneal commented:
Removed the logging and changed time so we get events every 15 minutes.  Process should now be "Wonder why that node isn't consolidated ->  `kubectl describe node node-name`" and observe events indicating why.
